File: frontend/app/shared/types/problem-details.ts
/**
 * Giao diện lỗi Problem Details chuẩn RFC 9457 mở rộng từ Backend Spring Boot.
 */
export interface ProblemDetails {
  type?: string;
  title: string;
  status: number;
  detail: string;
  instance?: string;
  code?: string;
  traceId?: string;
  errors?: Array<{
    field: string;
    code: string;
    message: string;
  }>;
}

export class ApiProblemException extends Error {
  problem: ProblemDetails;

  constructor(problem: ProblemDetails) {
    super(problem.detail || problem.title);
    this.name = 'ApiProblemException';
    this.problem = problem;
  }
}

File: backend/src/main/java/com/company/bds/shared/error/ProblemDetails.java
package com.company.bds.shared.error;

import com.fasterxml.jackson.annotation.JsonInclude;

import java.net.URI;
import java.util.List;

/**
 * Cấu trúc phản hồi lỗi API chuẩn hóa theo RFC 9457 mở rộng.
 * Tham khảo: PROJECT_CODE_RULES_BDS.md mục 8.4
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public record ProblemDetails(
        URI type,
        String title,
        int status,
        String detail,
        String instance,
        String code,
        String traceId,
        List<ValidationErrorItem> errors
) {
    public record ValidationErrorItem(
            String field,
            String code,
            String message
    ) {}
}
